### ISSUE TITLE
Simplify cubeb_register_device_collection_changed?

### ISSUE THREAD

**Issue by kinetiknz** [kinetiknz/cubeb#516]
cubeb_register_device_collection_changed seems over complicated, but it's possible I don't understand the reason for it's current design.

As of now, the complexity stems from the ability to register different callbacks for INPUT and OUTPUT via the same API.  The API also supports INPUT|OUTPUT, in which case the same callback is used for both types.

Looking at the history in BMO #1498242, I think we added support for separate INPUT and OUTPUT callbacks to allow the caller to define different behaviour for each device type.

The supplied callback is of the form: ` void cubeb_device_collection_changed_callback(cubeb * context, void * user_ptr);`.

It seems like this API could be simplified by supporting only a single callback registration, but changing the callback type to `void cubeb_device_collection_changed_callback(cubeb * context, void * user_ptr, cubeb_device_type devtype);` to allow the implementation to indicate which device type collection has changed.

Looking at the single existing use of the API in Gecko's CubebDeviceEnumerator, the input and output callbacks both call into CubebDeviceEnumerator::AudioDeviceListChanged and pass INPUT or OUTPUT depending on which cubeb_device_collection_changed_callback originated the call.

@achronop (and anyone else that might have an opinion) thoughts?

**Comment by achronop** [comment]
The enhancement you are describing is fair and it can be adapted easily in CubebDeviceEnumerator. One problem is the API change. Since cubeb hasn't had many users this is not a big issue. Another issue is that we must update all the backends on different platforms which is always cumbersome.

Other than that, I would prefer that `cubeb_device_type`, in the device callback, will be allowed to be either INPUT or OUTPUT, never INPUT|OUTPUT. I think this is what most platforms support but it would be better if we assert for that.

Overall, I think we should go on and implement it. Simplifying, cubeb makes our life easier. Can you please describe how you plan to implement the changes? Can I help to update the backends?

**Comment by kinetiknz** [comment]
> Other than that, I would prefer that cubeb_device_type, in the device callback, will be allowed to be either INPUT or OUTPUT, never INPUT|OUTPUT.

That sounds like a reasonable request.  Is it for symmetry with cubeb_enumerate_devices? Or do you mind explaining why?

In terms of handling the API change, via a flag day per https://github.com/kinetiknz/cubeb/issues/518#issuecomment-510006191.

**Comment by achronop** [comment]
> Or do you mind explaining why?

I believe it's more straight forward for the client. Also, the `INPUT | OUTPUT` type requires an ugly cast. 
On top of that, I would prefer if cubeb guarantees that the callback is referring to one device id. 

With input|output is not clear what the behavior is. For example, a device id has input and output side what do you expect, separated input-output callback or one input|output? If 2 device-ids, one input, and one output are removed at once what do you expect one or two callbacks? I would prefer to expect 2 callbacks in both cases.

Finally, from the implementation point of view is easier to implement the separated callback approach rather than the other way around. Especially if the platform does not support `input|output` notifications. If we continue supporting input|output we will end up having input|output callback wherever the underline platform supports it, and for the same case on another platform, we will have separated callbacks. I think that what we are doing but I have to duple check that. 

and yes it's also symmetric to `cubeb_enumerate_devices`.

**Comment by kinetiknz** [comment]
>  On top of that, I would prefer if cubeb guarantees that the callback is referring to one device id.

The current API doesn't do that as far as I understand it.  There's no device id specified in the registration or the callback.

My mental model for this API is that it's used to invalidate any cached device ids (i.e. it's invalidating the entire enumeration result) from cubeb_enumerate_devices - i.e. once you receive a device_collection_changed, you should re-enumerate and act accordingly.

**Comment by achronop** [comment]
No, the current API does not do that and it is not something that we really need. We use that API to invalidate the cached device list and re-enumerate similar to what you are describing.
